serve/mlc_serve/model/tvm_model.py

```diff
@@ -398,12 +454,22 @@ def generate(
             self.copy_cache_blocks_func(self.cache_blocks, block_mapping)
             cache.pending_copy_from_to = []
 
-        return sample_from_logits(logits, sequence_ids, requests, self.vocab_size)
+        return sample_from_logits(
+            logits,
+            sequence_ids,
+            requests,
+            sampling_metadata,
+            self.vocab_size,
+            self._copy_stream,
+            self.torch_dtype,
+            self.torch_dev,
+            past_decode_tokens,
+        )
 
 
 def init_tvm_model(
     model_artifact_config: ModelArtifactConfig, engine_config: MLCServeEngineConfig
-) -> Tuple[TextGenerator, CacheManager]:
+) -> Tuple[Model, CacheManager]:
```

Comment:
In my PT integration PR, I'm adding `init_torch_model` function, which also has the return type `TextGenerator`. This way, `PagedCacheModelTextGenerator` can work with both `Model` implemented in `tvm_model.py` and `torch_model.py` like this:

```
if engine_config.model_type == "tvm":
    model_artifact_config = get_model_artifact_config(model_artifact_path)
    model, cache_manager = init_tvm_model(model_artifact_config, engine_config)
    tokenizer_module = HfTokenizerModule(model_artifact_path.joinpath("model"))
elif engine_config.model_type == "torch":
    model, cache_manager, model_artifact_config = init_torch_model(
        model_artifact_path, engine_config
    )
    tokenizer_module = HfTokenizerModule(model_artifact_path)

self.text_generator = PagedCacheModelTextGenerator(model)
``` 

This is why we want the return type be a protocol, `TextGenerator`, not a concrete class, `Model`. Also cc @vvchernov

Reply:
Ok, I understand the idea, why do we not use inheritance and `TextGenerator` as base class?

Reply:
We could have defined `TextGenerator` as an abstract base class and define `Model` as a subclass. Protocol is an alternative way for defining subtyping relations, which is supposed to be more flexible since it doesn't require a hierarchy. We chose Protocol-based subtyping, so I would say subclassing from `TextGenerator` defeats the purpose of such subtyping. See more in https://mypy.readthedocs.io/en/stable/protocols.html

Reply:
The `protocol` is probably still the better way so that we can interface with vLLM and other things where we may not have the full control of the hierarchy.